File: src/client/asmlibs.rs
#![crate_name = "asmlibs"]
#![no_std]

#![feature(no_std)]
#![feature(core, core_prelude, core_slice_ext)]
#![feature(raw)]

extern crate core;
extern crate physics;
extern crate graphics;
#[macro_use] extern crate asmrt;

use core::prelude::*;
use core::mem;
use core::raw;
use core::slice;
use physics::v3::{V3, V2, scalar, Region};
use physics::{Shape, ShapeSource};
use physics::{CHUNK_SIZE, CHUNK_BITS, CHUNK_MASK, TILE_BITS};

use graphics::lights;
use graphics::structures;
use graphics::terrain;
use graphics::types as gfx_types;

mod std {
    pub use core::fmt;
    pub use core::marker;
}


pub const LOCAL_BITS: usize = 3;
pub const LOCAL_SIZE: i32 = 1 << LOCAL_BITS;    // 8
pub const LOCAL_MASK: i32 = LOCAL_SIZE - 1;

pub const REPEAT_BITS: i32 = 1;
pub const REPEAT_SIZE: i32 = 1 << REPEAT_BITS;  // 2
pub const REPEAT_MASK: i32 = REPEAT_SIZE - 1;

pub const NUM_LAYERS: usize = 3;


// Physics

pub type ShapeChunk = [Shape; 1 << (3 * CHUNK_BITS)];

pub struct ShapeLayers {
    base: ShapeChunk,
    layers: [ShapeChunk; NUM_LAYERS],
    merged: ShapeChunk,
}

impl ShapeLayers {
    fn refresh(&mut self, bounds: Region) {
        let chunk_bounds = Region::new(scalar(0), scalar(CHUNK_SIZE));

        for p in bounds.intersect(chunk_bounds).points() {
            let idx = chunk_bounds.index(p);
            self.merged[idx] = self.base[idx];

            for layer in self.layers.iter() {
                if shape_overrides(self.merged[idx], layer[idx]) {
                    self.merged[idx] = layer[idx];
                }
            }
        }
    }
}

fn shape_overrides(old: Shape, new: Shape) -> bool {
    match (old, new) {
        (Shape::Empty, _) => true,

        (Shape::Floor, Shape::Empty) => false,
        (Shape::Floor, _) => true,

        (Shape::Solid, _) => false,

        _ => false,
    }
}

struct AsmJsShapeSource<'a> {
    layers: &'a [ShapeLayers; 1 << (2 * LOCAL_BITS)],
}

impl<'a> ShapeSource for AsmJsShapeSource<'a> {
    fn get_shape(&self, pos: V3) -> Shape {
        if pos.z < 0 || pos.z >= CHUNK_SIZE {
            return Shape::Empty;
        }

        let V3 { x: tile_x, y: tile_y, z: tile_z } = pos & V3::new(CHUNK_MASK, CHUNK_MASK, -1);
        let V3 { x: chunk_x, y: chunk_y, z: _ } = (pos >> CHUNK_BITS) & scalar(LOCAL_MASK);

        let chunk_idx = chunk_y * LOCAL_SIZE + chunk_x;
        let tile_idx = (tile_z * CHUNK_SIZE + tile_y) * CHUNK_SIZE + tile_x;

        let shape = self.layers[chunk_idx as usize].merged[tile_idx as usize];
        shape
    }
}


#[derive(Clone, Copy)]
pub struct CollideArgs {
    pub pos: V3,
    pub size: V3,
    pub velocity: V3,
}

#[derive(Clone, Copy)]
pub struct CollideResult {
    pub pos: V3,
    pub time: i32,
}

#[export_name = "collide"]
pub extern fn collide_wrapper(layers: &[ShapeLayers; 1 << (2 * LOCAL_BITS)],
                              input: &CollideArgs,
                              output: &mut CollideResult) {
    let (pos, time) = physics::collide(&AsmJsShapeSource { layers: layers },
                                       input.pos, input.size, input.velocity);
    output.pos = pos;
    output.time = time;
}

#[export_name = "set_region_shape"]
pub extern fn set_region_shape(layers: &mut [ShapeLayers; 1 << (2 * LOCAL_BITS)],
                               bounds: &Region,
                               layer: usize,
                               shape_data: *const Shape,
                               shape_len: usize) {
    let shape: &[Shape] = unsafe {
        mem::transmute(raw::Slice {
            data: shape_data,
            len: shape_len,
        })
    };

    let chunk_bounds = Region::new(scalar(0), scalar(CHUNK_SIZE));
    for p in bounds.points() {
        // div_floor requires an extra LLVM intrinsic.
        let cpos = p.reduce() >> CHUNK_BITS;
        let masked_cpos = cpos & scalar(LOCAL_MASK);
        let cidx = masked_cpos.y * LOCAL_SIZE + masked_cpos.x;

        let offset = p & scalar(CHUNK_MASK);
        let out_idx = chunk_bounds.index(offset);
        let in_idx = bounds.index(p);

        layers[cidx as usize].layers[layer][out_idx] = shape[in_idx];
    }
}

#[export_name = "refresh_shape_cache"]
pub extern fn refresh_shape_cache(layers: &mut [ShapeLayers; 1 << (2 * LOCAL_BITS)],
                                  bounds: &Region) {
    let chunk_bounds = bounds.reduce().div_round(CHUNK_SIZE);

    for cpos in chunk_bounds.points() {
        let masked_cpos = cpos & scalar(LOCAL_MASK);
        let cidx = masked_cpos.y * LOCAL_SIZE + masked_cpos.x;

        let base = cpos.extend(0) * scalar(CHUNK_SIZE);
        layers[cidx as usize].refresh(*bounds - base);
    }
}

#[export_name = "find_ceiling"]
pub extern fn find_ceiling(layers: &[ShapeLayers; 1 << (2 * LOCAL_BITS)],
                           pos: &V3) -> i32 {
    let tpos = *pos >> TILE_BITS;

    let cx = tpos.x / CHUNK_SIZE % LOCAL_SIZE;
    let cy = tpos.y / CHUNK_SIZE % LOCAL_SIZE;
    let idx = (cy * LOCAL_SIZE + cx) as usize;

    let x = tpos.x % CHUNK_SIZE;
    let y = tpos.y % CHUNK_SIZE;
    let mut z = tpos.z + 1;

    let chunk = &layers[idx];
    while z < 16 {
        let tile_idx = x + CHUNK_SIZE * (y + CHUNK_SIZE * (z));
        if chunk.merged[tile_idx as usize] != Shape::Empty {
            break;
        }
        z += 1;
    }
    z
}


// Graphics

unsafe fn make_slice<T>(ptr: *const T, byte_len: usize) -> &'static [T] {
    slice::from_raw_parts(ptr, byte_len / mem::size_of::<T>())
}

unsafe fn make_slice_mut<T>(ptr: *mut T, byte_len: usize) -> &'static mut [T] {
    slice::from_raw_parts_mut(ptr, byte_len / mem::size_of::<T>())
}

pub struct GeometryResult {
    vertex_count: usize,
    more: u8,
}


#[export_name = "terrain_geom_init"]
pub unsafe extern fn terrain_geom_init(geom: &mut terrain::GeomGen<'static>,
                                       block_data_ptr: *const gfx_types::BlockData,
                                       block_data_byte_len: usize,
                                       local_chunks: &'static gfx_types::LocalChunks) {
    let block_data = make_slice(block_data_ptr, block_data_byte_len);
    geom.init(block_data, local_chunks);
}

#[export_name = "terrain_geom_reset"]
pub extern fn terrain_geom_reset(geom: &mut terrain::GeomGen,
                                 cx: i32,
                                 cy: i32) {
    geom.reset(V2::new(cx, cy));
}

#[export_name = "terrain_geom_generate"]
pub unsafe extern fn terrain_geom_generate(geom: &mut terrain::GeomGen,
                                           buf_ptr: *mut terrain::Vertex,
                                           buf_byte_len: usize,
                                           result: &mut GeometryResult) {
    let buf = make_slice_mut(buf_ptr, buf_byte_len);

    let mut idx = 0;
    let more = geom.generate(buf, &mut idx);

    result.vertex_count = idx;
    result.more = more as u8;
}


#[export_name = "structure_buffer_init"]
pub unsafe extern fn structure_buffer_init(buf: &mut structures::Buffer<'static>,
                                           storage_ptr: *mut structures::Structure,
                                           storage_byte_len: usize) {
    let storage = make_slice_mut(storage_ptr, storage_byte_len);
    buf.init(storage);
}

#[export_name = "structure_buffer_insert"]
pub extern fn structure_buffer_insert(buf: &mut structures::Buffer,
                                      external_id: u32,
                                      pos_x: u8,
                                      pos_y: u8,
                                      pos_z: u8,
                                      template_id: u32,
                                      oneshot_start: u16) -> usize {
    if let Some(idx) = buf.insert(external_id,
                                  (pos_x, pos_y, pos_z),
                                  template_id) {
        buf[idx].oneshot_start = oneshot_start;
        idx
    } else {
        (-1_isize) as usize
    }
}

#[export_name = "structure_buffer_remove"]
pub extern fn structure_buffer_remove(buf: &mut structures::Buffer,
                                      idx: usize) -> u32 {
    buf.remove(idx)
}


#[export_name = "structure_geom_init"]
pub unsafe extern fn structure_geom_init(geom: &mut structures::GeomGen<'static>,
                                         buffer: &'static structures::Buffer<'static>,
                                         templates_ptr: *const gfx_types::StructureTemplate,
                                         templates_byte_len: usize,
                                         parts_ptr: *const gfx_types::TemplatePart,
                                         parts_byte_len: usize,
                                         verts_ptr: *const gfx_types::TemplateVertex,
                                         verts_byte_len: usize) {
    let templates = make_slice(templates_ptr, templates_byte_len);
    let parts = make_slice(parts_ptr, parts_byte_len);
    let verts = make_slice(verts_ptr, verts_byte_len);
    geom.init(buffer, templates, parts, verts);
}

#[export_name = "structure_geom_reset"]
pub extern fn structure_geom_reset(geom: &mut structures::GeomGen,
                                   cx0: i32,
                                   cy0: i32,
                                   cx1: i32,
                                   cy1: i32,
                                   sheet: u8) {
    geom.reset(Region::new(V2::new(cx0, cy0),
                           V2::new(cx1, cy1)),
               sheet);
}

#[export_name = "structure_geom_generate"]
pub unsafe extern fn structure_geom_generate(geom: &mut structures::GeomGen,
                                             buf_ptr: *mut structures::Vertex,
                                             buf_byte_len: usize,
                                             result: &mut GeometryResult) {
    let buf = make_slice_mut(buf_ptr, buf_byte_len);

    let mut idx = 0;
    let more = geom.generate(buf, &mut idx);

    result.vertex_count = idx;
    result.more = more as u8;
}


#[export_name = "light_geom_init"]
pub unsafe extern fn light_geom_init(geom: &mut lights::GeomGen<'static>,
                                     buffer: &'static structures::Buffer<'static>,
                                     templates_ptr: *const gfx_types::StructureTemplate,
                                     templates_byte_len: usize) {
    let templates = make_slice(templates_ptr, templates_byte_len);
    geom.init(buffer, templates);
}

#[export_name = "light_geom_reset"]
pub extern fn light_geom_reset(geom: &mut lights::GeomGen,
                               cx0: i32,
                               cy0: i32,
                               cx1: i32,
                               cy1: i32) {
    geom.reset(Region::new(V2::new(cx0, cy0),
                           V2::new(cx1, cy1)));
}

#[export_name = "light_geom_generate"]
pub unsafe extern fn light_geom_generate(geom: &mut lights::GeomGen,
                                         buf_ptr: *mut lights::Vertex,
                                         buf_byte_len: usize,
                                         result: &mut GeometryResult) {
    let buf = make_slice_mut(buf_ptr, buf_byte_len);

    let mut idx = 0;
    let more = geom.generate(buf, &mut idx);

    result.vertex_count = idx;
    result.more = more as u8;
}


// SIZEOF

#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct Sizes {
    shape_chunk: usize,
    shape_layers: usize,

    block_data: usize,
    block_chunk: usize,
    local_chunks: usize,

    structure: usize,

    terrain_vertex: usize,
    terrain_geom_gen: usize,

    structures_template: usize,
    template_part: usize,
    template_vertex: usize,
    structures_buffer: usize,
    structures_vertex: usize,
    structures_geom_gen: usize,

    light_vertex: usize,
    light_geom_gen: usize,
}

#[export_name = "get_sizes"]
pub extern fn get_sizes(sizes: &mut Sizes, num_sizes: &mut usize) {
    use core::mem::size_of;

    sizes.shape_chunk = size_of::<ShapeChunk>();
    sizes.shape_layers = size_of::<ShapeLayers>();

    sizes.block_data = size_of::<gfx_types::BlockData>();
    sizes.block_chunk = size_of::<gfx_types::BlockChunk>();
    sizes.local_chunks = size_of::<gfx_types::LocalChunks>();

    sizes.structure = size_of::<structures::Structure>();

    sizes.terrain_vertex = size_of::<terrain::Vertex>();
    sizes.terrain_geom_gen = size_of::<terrain::GeomGen>();

    sizes.structures_template = size_of::<gfx_types::StructureTemplate>();
    sizes.template_part = size_of::<gfx_types::TemplatePart>();
    sizes.template_vertex = size_of::<gfx_types::TemplateVertex>();
    sizes.structures_buffer = size_of::<structures::Buffer>();
    sizes.structures_vertex = size_of::<structures::Vertex>();
    sizes.structures_geom_gen = size_of::<structures::GeomGen>();

    sizes.light_vertex = size_of::<lights::Vertex>();
    sizes.light_geom_gen = size_of::<lights::GeomGen>();

    *num_sizes = size_of::<Sizes>() / size_of::<usize>();
}


#[export_name = "test"]
pub extern fn test_wrapper() {
}

File: src/client/js/asmlibs.js
var module = window['asmlibs_code'];
var static_data = window['asmlibs_data'];

var Vec = require('util/vec').Vec;
var decodeUtf8 = require('util/misc').decodeUtf8;
var LOCAL_SIZE = require('data/chunk').LOCAL_SIZE;


// Memory layout

// Emscripten puts the first static at address 8 to  avoid storing data at
// address 0.
var STATIC_START = 8;  
var STATIC_SIZE = static_data.byteLength;
var STATIC_END = STATIC_START + STATIC_SIZE;

// Align STACK_START to an 8-byte boundary.
var STACK_START = (STATIC_END + 7) & ~7;
// Give at least 4k for stack, and align to 4k boundary.
var STACK_END = (STACK_START + 0x1000 + 0x0fff) & ~0x0fff;
var STACK_SIZE = STACK_END - STACK_START;
console.assert(STACK_SIZE >= 0x1000, 'want at least 4kb for stack');

var HEAP_START = STACK_END;


// External functions

var msg_buffer = '';

var module_env = function(buffer, callback_array) {
    return ({
        'abort': function() {
            console.assert(false, 'abort');
            throw 'abort';
        },

        'traceInts': function(ptr, len) {
            var arr = [];
            var view = new Int32Array(buffer);
            for (var i = 0; i < len; ++i) {
                arr.push(view[(ptr >> 2) + i]);
            }
            console.log(arr);
        },

        'physTrace': function(x, y, w, h) {
            window.physTrace.push([x, y, w, h]);
        },

        'resetPhysTrace': function() {
            window.physTrace = [];
        },

        'writeStr': function(ptr, len) {
            var view = new Uint8Array(buffer, ptr, len);
            msg_buffer += decodeUtf8(view);
        },

        'flushStr': function() {
            console.log(msg_buffer);
            msg_buffer = '';
        },

        'runCallback': function(index, arg_base, arg_len) {
            var cb = callback_array[index];
            console.assert(cb != null, 'bad callback index: ', index);

            var args = new Int32Array(buffer, arg_base, arg_len);
            return cb.apply(window, args);
        },

        'STACK_START': STACK_START,
        'STACK_END': STACK_END,
    });
};


// Helper functions

function memcpy(dest_buffer, dest_offset, src_buffer, src_offset, len) {
    var dest = new Int8Array(dest_buffer, dest_offset, len);
    var src = new Int8Array(src_buffer, src_offset, len);
    dest.set(src);
}


// Buffer size computations
var SIZEOF = (function() {
    var buffer = new ArrayBuffer((HEAP_START + 0xffff) & ~0xffff);
    memcpy(buffer, STATIC_START,
            static_data.buffer, static_data.byteOffset, static_data.byteLength);
    var asm = module(window, module_env(buffer), buffer);

    var EXPECT_SIZES = 17;
    var alloc = ((1 + EXPECT_SIZES) * 4 + 7) & ~7;
    var base = asm['__adjust_stack'](alloc);

    asm['get_sizes'](base + 4, base);

    var view = new Int32Array(buffer, base, 1 + EXPECT_SIZES);
    console.assert(view[0] == EXPECT_SIZES,
            'expected sizes for ' + EXPECT_SIZES + ' types, but got ' + view[0]);

    var sizeof = {};
    var index = 0;
    var next = function() { return view[1 + index++]; };

    sizeof.ShapeChunk = next();
    sizeof.ShapeLayers = next();

    sizeof.BlockData = next();
    sizeof.BlockChunk = next();
    sizeof.LocalChunks = next();

    sizeof.Structure = next();

    sizeof.TerrainVertex = next();
    sizeof.TerrainGeomGen = next();

    sizeof.StructureTemplate = next();
    sizeof.ModelVertex = next();
    sizeof.StructureBuffer = next();
    sizeof.StructureBaseVertex = next();
    sizeof.StructureBaseGeomGen = next();
    sizeof.StructureAnimVertex = next();
    sizeof.StructureAnimGeomGen = next();

    sizeof.LightVertex = next();
    sizeof.LightGeomGen = next();

    console.assert(index == EXPECT_SIZES,
            'some items were left over after building sizeof', index, EXPECT_SIZES);

    return sizeof;
})();
exports.SIZEOF = SIZEOF;


// window.Asm wrapper

function next_heap_size(size) {
    // "the heap object's byteLength must be either 2^n for n in [12, 24) or
    // 2^24 * n for n ≥ 1"
    if (size <= (1 << 12)) {
        return (1 << 12);
    } else if (size >= (1 << 24)) {
        return (size | ((1 << 24) - 1)) + 1;
    } else {
        for (var i = 12 + 1; i < 24; ++i) {
            if (size <= (1 << i)) {
                return (1 << i);
            }
        }
        console.assert(false, 'failed to compute next heap size for', size);
        return (1 << 24);
    }
}

/** @constructor */
function Asm(heap_size) {
    // Buffer size must be a multiple of 4k.
    var min_size = HEAP_START + heap_size;
    this.buffer = new ArrayBuffer(next_heap_size(min_size));

    this._callbacks = [];

    this.memcpy(STATIC_START, static_data);
    this._raw = module(window, module_env(this.buffer, this._callbacks), this.buffer);
}
exports.Asm = Asm;

Asm.prototype._stackAlloc = function(type, count) {
    var size = count * type.BYTES_PER_ELEMENT;
    var base = this._raw['__adjust_stack']((size + 7) & ~7);
    return new type(this.buffer, base, count);
};

Asm.prototype._stackFree = function(view) {
    var size = view.byteLength;
    this._raw['__adjust_stack'](-((size + 7) & ~7));
};

Asm.prototype._callbackAlloc = function(cb) {
    this._callbacks.push(cb);
    return this._callbacks.length - 1;
};

Asm.prototype._callbackFree = function(idx) {
    console.assert(idx == this._callbacks.length - 1);
    this._callbacks.pop();
};

Asm.prototype._storeVec = function(view, offset, v) {
    view[offset + 0] = v.x;
    view[offset + 1] = v.y;
    view[offset + 2] = v.z;
};

Asm.prototype._makeView = function(type, offset, bytes) {
    return new type(this.buffer, offset, bytes / type.BYTES_PER_ELEMENT);
};

Asm.prototype.memcpy = function(dest_offset, data) {
    memcpy(this.buffer, dest_offset, data.buffer, data.byteOffset, data.byteLength);
};


// Physics

var PHYSICS_HEAP_START = HEAP_START;

var SHAPE_LAYERS_START = HEAP_START;
var SHAPE_LAYERS_END = SHAPE_LAYERS_START + SIZEOF.ShapeLayers * LOCAL_SIZE * LOCAL_SIZE;

var PHYSICS_HEAP_END = SHAPE_LAYERS_END;

Asm.prototype.collide = function(pos, size, velocity) {
    var input = this._stackAlloc(Int32Array, 9);
    var output = this._stackAlloc(Int32Array, 4);

    this._storeVec(input, 0, pos);
    this._storeVec(input, 3, size);
    this._storeVec(input, 6, velocity);

    this._raw['collide'](SHAPE_LAYERS_START, input.byteOffset, output.byteOffset);

    var result = ({
        x: output[0],
        y: output[1],
        z: output[2],
        t: output[3],
    });

    this._stackFree(output);
    this._stackFree(input);

    return result;
};

Asm.prototype.setRegionShape = function(pos, size, layer, shape) {
    var input_bounds = this._stackAlloc(Int32Array, 6);
    var input_shape = this._stackAlloc(Uint8Array, shape.length);

    this._storeVec(input_bounds, 0, pos);
    this._storeVec(input_bounds, 3, pos.add(size));
    input_shape.set(shape);

    this._raw['set_region_shape'](SHAPE_LAYERS_START,
            input_bounds.byteOffset, layer,
            input_shape.byteOffset, input_shape.length);
    this._raw['refresh_shape_cache'](SHAPE_LAYERS_START, input_bounds.byteOffset);

    this._stackFree(input_shape);
    this._stackFree(input_bounds);
};

Asm.prototype.clearRegionShape = function(pos, size, layer) {
    var volume = size.x * size.y * size.z;

    var input_bounds = this._stackAlloc(Int32Array, 6);
    var input_shape = this._stackAlloc(Uint8Array, volume);

    this._storeVec(input_bounds, 0, pos);
    this._storeVec(input_bounds, 3, pos.add(size));
    for (var i = 0; i < volume; ++i) {
        input_shape[i] = 0;
    }

    this._raw['set_region_shape'](PHYSICS_HEAP_START,
            input_bounds.byteOffset, layer,
            input_shape.byteOffset, input_shape.length);
    this._raw['refresh_shape_cache'](PHYSICS_HEAP_START, input_bounds.byteOffset);

    this._stackFree(input_shape);
    this._stackFree(input_bounds);
};

Asm.prototype.refreshShapeLayers = function(pos, size) {
    var input = this._stackAlloc(Int32Array, 6);

    this._storeVec(input, 0, pos);
    this._storeVec(input, 3, pos.add(size));

    this._raw['refresh_shape_cache'](SHAPE_LAYERS_START, input.byteOffset);

    this._stackFree(input);
};

Asm.prototype.shapeLayerView = function(chunk_idx, layer) {
    var chunk_offset = chunk_idx * SIZEOF.ShapeLayers;
    var layer_offset = (1 + layer) * SIZEOF.ShapeChunk;

    return new Uint8Array(this.buffer,
            SHAPE_LAYERS_START + chunk_offset + layer_offset, SIZEOF.ShapeChunk);
};

Asm.prototype.findCeiling = function(pos) {
    var input = this._stackAlloc(Int32Array, 3);
    this._storeVec(input, 0, pos);

    var result = this._raw['find_ceiling'](SHAPE_LAYERS_START, input.byteOffset);

    this._stackFree(input);
    return result;
};

exports.getPhysicsHeapSize = function() {
    return PHYSICS_HEAP_END - PHYSICS_HEAP_START;
};


// Graphics

/** @constructor */
function AsmGraphics(num_blocks, num_templates, num_model_vertexes,
        structures_size, geom_size) {
    var heap_end = HEAP_START;
    function alloc(size) {
        // 8-byte alignment
        heap_end = (heap_end + 7) & ~7;
        var pos = heap_end;
        heap_end += size;
        return pos;
    }

    this.num_blocks = num_blocks;
    this.num_templates = num_templates;
    this.num_model_vertexes = num_model_vertexes;

    this.block_data_bytes = num_blocks * SIZEOF.BlockData;
    this.template_data_bytes = num_templates * SIZEOF.StructureTemplate;
    this.model_vertex_bytes = num_model_vertexes * SIZEOF.ModelVertex;
    this.geom_buffer_bytes = geom_size;
    // TODO: sizeof(Structure) * num_structures
    this.structure_storage_bytes = structures_size

    this.LOCAL_CHUNKS = alloc(SIZEOF.LocalChunks);
    this.TERRAIN_GEOM_GEN = alloc(SIZEOF.TerrainGeomGen);
    this.STRUCTURE_BASE_GEOM_GEN = alloc(SIZEOF.StructureBaseGeomGen);
    this.STRUCTURE_ANIM_GEOM_GEN = alloc(SIZEOF.StructureAnimGeomGen);
    this.LIGHT_GEOM_GEN = alloc(SIZEOF.LightGeomGen);
    this.STRUCTURE_BUFFER = alloc(SIZEOF.StructureBuffer);

    this.BLOCK_DATA = alloc(this.block_data_bytes);
    this.TEMPLATE_DATA = alloc(this.template_data_bytes);
    this.MODEL_DATA = alloc(this.model_vertex_bytes);
    this.GEOM_BUFFER = alloc(this.geom_buffer_bytes);

    this.STRUCTURE_STORAGE = alloc(this.structure_storage_bytes);

    Asm.call(this, heap_end - HEAP_START);
}
AsmGraphics.prototype = Object.create(Asm.prototype);
exports.AsmGraphics = AsmGraphics;


AsmGraphics.prototype.blockDataView8 = function() {
    return this._makeView(Uint8Array, this.BLOCK_DATA, this.block_data_bytes);
};

AsmGraphics.prototype.blockDataView16 = function() {
    return this._makeView(Uint16Array, this.BLOCK_DATA, this.block_data_bytes);
};

AsmGraphics.prototype.chunkView = function(cx, cy) {
    var idx = (cy & (LOCAL_SIZE - 1)) * LOCAL_SIZE + (cx & (LOCAL_SIZE - 1));
    var offset = idx * SIZEOF.BlockChunk;
    return this._makeView(Uint16Array, this.LOCAL_CHUNKS + offset, SIZEOF.BlockChunk);
};

AsmGraphics.prototype.templateDataView8 = function() {
    return this._makeView(Uint8Array, this.TEMPLATE_DATA, this.template_data_bytes);
};

AsmGraphics.prototype.templateDataView16 = function() {
    return this._makeView(Uint16Array, this.TEMPLATE_DATA, this.template_data_bytes);
};

AsmGraphics.prototype.modelVertexView = function() {
    return this._makeView(Uint16Array, this.MODEL_DATA, this.model_vertex_bytes);
};


AsmGraphics.prototype.terrainGeomInit = function() {
    this._raw['terrain_geom_init'](
            this.TERRAIN_GEOM_GEN,
            this.BLOCK_DATA,
            this.block_data_bytes,
            this.LOCAL_CHUNKS);
};

AsmGraphics.prototype.terrainGeomReset = function(cx, cy) {
    this._raw['terrain_geom_reset'](this.TERRAIN_GEOM_GEN, cx, cy);
};

AsmGraphics.prototype.terrainGeomGenerate = function() {
    var output = this._stackAlloc(Int32Array, 2);

    this._raw['terrain_geom_generate'](
            this.TERRAIN_GEOM_GEN,
            this.GEOM_BUFFER,
            this.geom_buffer_bytes,
            output.byteOffset);

    var vertex_count = output[0];
    var more = (output[1] & 1) != 0;

    this._stackFree(output);

    return {
        geometry: this._makeView(Uint8Array, this.GEOM_BUFFER,
                          vertex_count * SIZEOF.TerrainVertex),
        more: more,
    };
};


AsmGraphics.prototype.structureBufferInit = function() {
    this._raw['structure_buffer_init'](
            this.STRUCTURE_BUFFER,
            this.STRUCTURE_STORAGE,
            this.structure_storage_bytes);
};

AsmGraphics.prototype.structureBufferInsert = function(id, x, y, z, template_id) {
    return this._raw['structure_buffer_insert'](
            this.STRUCTURE_BUFFER,
            id, x, y, z, template_id);
};

AsmGraphics.prototype.structureBufferRemove = function(idx) {
    return this._raw['structure_buffer_remove'](
            this.STRUCTURE_BUFFER,
            idx);
};

AsmGraphics.prototype.structureBufferSetOneshotStart = function(idx, oneshot_start) {
    this._raw['structure_buffer_set_oneshot_start'](
            this.STRUCTURE_BUFFER,
            idx, oneshot_start);
};


AsmGraphics.prototype.structureBaseGeomInit = function() {
    this._raw['structure_base_geom_init'](
            this.STRUCTURE_BASE_GEOM_GEN,
            this.STRUCTURE_BUFFER,
            this.TEMPLATE_DATA,
            this.template_data_bytes,
            this.MODEL_DATA,
            this.model_vertex_bytes);
};

AsmGraphics.prototype.structureBaseGeomReset = function(cx0, cy0, cx1, cy1, sheet) {
    this._raw['structure_base_geom_reset'](
            this.STRUCTURE_BASE_GEOM_GEN,
            cx0, cy0, cx1, cy1, sheet);
};

AsmGraphics.prototype.structureBaseGeomGenerate = function() {
    var output = this._stackAlloc(Int32Array, 2);

    this._raw['structure_base_geom_generate'](
            this.STRUCTURE_BASE_GEOM_GEN,
            this.GEOM_BUFFER,
            this.geom_buffer_bytes,
            output.byteOffset);

    var vertex_count = output[0];
    var more = (output[1] & 1) != 0;

    this._stackFree(output);

    return {
        geometry: this._makeView(Uint8Array, this.GEOM_BUFFER,
                          vertex_count * SIZEOF.StructureBaseVertex),
        more: more,
    };
};


AsmGraphics.prototype.structureAnimGeomInit = function() {
    this._raw['structure_anim_geom_init'](
            this.STRUCTURE_ANIM_GEOM_GEN,
            this.STRUCTURE_BUFFER,
            this.TEMPLATE_DATA,
            this.template_data_bytes);
};

AsmGraphics.prototype.structureAnimGeomReset = function(cx0, cy0, cx1, cy1, sheet) {
    this._raw['structure_anim_geom_reset'](
            this.STRUCTURE_ANIM_GEOM_GEN,
            cx0, cy0, cx1, cy1, sheet);
};

AsmGraphics.prototype.structureAnimGeomGenerate = function() {
    var output = this._stackAlloc(Int32Array, 2);

    this._raw['structure_anim_geom_generate'](
            this.STRUCTURE_ANIM_GEOM_GEN,
            this.GEOM_BUFFER,
            this.geom_buffer_bytes,
            output.byteOffset);

    var vertex_count = output[0];
    var more = (output[1] & 1) != 0;

    this._stackFree(output);

    return {
        geometry: this._makeView(Uint8Array, this.GEOM_BUFFER,
                          vertex_count * SIZEOF.StructureAnimVertex),
        more: more,
    };
};


AsmGraphics.prototype.lightGeomInit = function() {
    this._raw['light_geom_init'](
            this.LIGHT_GEOM_GEN,
            this.STRUCTURE_BUFFER,
            this.TEMPLATE_DATA,
            this.template_data_bytes);
};

AsmGraphics.prototype.lightGeomReset = function(cx0, cy0, cx1, cy1) {
    this._raw['light_geom_reset'](
            this.LIGHT_GEOM_GEN,
            cx0, cy0, cx1, cy1);
};

AsmGraphics.prototype.lightGeomGenerate = function() {
    var output = this._stackAlloc(Int32Array, 2);

    this._raw['light_geom_generate'](
            this.LIGHT_GEOM_GEN,
            this.GEOM_BUFFER,
            this.geom_buffer_bytes,
            output.byteOffset);

    var vertex_count = output[0];
    var more = (output[1] & 1) != 0;

    this._stackFree(output);

    return {
        geometry: this._makeView(Uint8Array, this.GEOM_BUFFER,
                          vertex_count * SIZEOF.LightVertex),
        more: more,
    };
};




// Test

Asm.prototype.test = function() {
};
